src/algebra/module/injective.lean

```diff
@@ -0,0 +1,591 @@
+begin
+  have mem1 : x.1 ∈ (_ : set _) := x.2,
+  rw submodule.coe_sup at mem1,
+  rcases mem1 with ⟨a, b, a_mem, (b_mem : b ∈ (submodule.span R _ : submodule R N)), eq1⟩,
+  rw submodule.mem_span_singleton at b_mem,
+  rcases b_mem with ⟨z, eq2⟩,
+  exact ⟨⟨a, a_mem⟩, z, by rw [← eq1, ← eq2]⟩,
+end
+
+/--If `x ∈ M ⊔ ⟨y⟩`, then `x = m + r • y`, `fst` pick an arbitrary such `m`.-/
+def extension_of_max_adjoin.fst
+  {y : N} (x : (extension_of_max hi f).to_submodule ⊔ submodule.span R {y}) :
+  (extension_of_max hi f).to_submodule :=
+(extension_of_max_adjoin.aux1 x).some
+
+/--If `x ∈ M ⊔ ⟨y⟩`, then `x = m + r • y`, `snd` pick an arbitrary such `r`.-/
+def extension_of_max_adjoin.snd
+  {y : N} (x : (extension_of_max hi f).to_submodule ⊔ submodule.span R {y}) :
+  R :=
+(extension_of_max_adjoin.aux1 x).some_spec.some
+
+lemma extension_of_max_adjoin.eqn
+  {y : N} (x : (extension_of_max hi f).to_submodule ⊔ submodule.span R {y}) :
+  x.1 = (extension_of_max_adjoin.fst x).1 + (extension_of_max_adjoin.snd x) • y :=
+(extension_of_max_adjoin.aux1 x).some_spec.some_spec
+
+variables (hi f)
+/--the ideal `I = {r | r • y ∈ N}`-/
+def extension_of_max_adjoin.ideal (y : N) :
+  ideal R :=
+{ carrier := {r | r • y ∈ (extension_of_max hi f).to_submodule},
+  add_mem' := λ a b (ha : a • y ∈ _) (hb : b • y ∈ _), begin
+    change (a + b) • _ ∈ _,
+    rw [add_smul],
+    exact submodule.add_mem _ ha hb,
+  end,
+  smul_mem' := λ c x (hx : x • y ∈ _), begin
+    change (_ • _) • _ ∈ _,
+    rw [smul_eq_mul, mul_smul],
+    exact submodule.smul_mem _ _ hx,
+  end,
+  zero_mem' := by simp }
+
+/--A linear map `I ⟶ Q` by `x ↦ f' (x • y)` where `f'` is the maximal extension-/
+def extension_of_max_adjoin.ideal_to (y : N) :
+  extension_of_max_adjoin.ideal hi f y →ₗ[R] Q :=
+{ to_fun := λ z, (extension_of_max hi f).extension ⟨z.1 • y, z.2⟩,
+  map_add' := λ z1 z2, by rw [← map_add]; congr; erw add_smul,
+  map_smul' := λ z1 z2, begin
+    rw [ring_hom.id_apply, ← linear_map.map_smul],
+    exact congr_arg _ (subtype.ext_val (mul_smul _ _ _)),
+  end }
+
+lemma extension_of_max_adjoin.aux2 (h : module.Baer R Q) (y : N) :
+  ∃ (g' : R →ₗ[R] Q), ∀ (x : R) (mem : x ∈ extension_of_max_adjoin.ideal hi f y),
+    g' x = extension_of_max_adjoin.ideal_to hi f y ⟨x, mem⟩ :=
+h (extension_of_max_adjoin.ideal hi f y) (extension_of_max_adjoin.ideal_to hi f y)
+
+/-- Since we assumed `Q` being Baer, the linear map `x ↦ f' (x • y) : I ⟶ Q` extends to `R ⟶ Q`,
+call this extended map `φ`-/
+def extension_of_max_adjoin.extend_ideal_to (h : module.Baer R Q) (y : N) : R →ₗ[R] Q :=
+(extension_of_max_adjoin.aux2 hi f h y).some
+
+lemma extension_of_max_adjoin.extend_ideal_to_is_extension (h : module.Baer R Q) (y : N) :
+  ∀ (x : R) (mem : x ∈ extension_of_max_adjoin.ideal hi f y),
+  extension_of_max_adjoin.extend_ideal_to hi f h y x =
+  extension_of_max_adjoin.ideal_to hi f y ⟨x, mem⟩ :=
+(extension_of_max_adjoin.aux2 hi f h y).some_spec
```

Comment:
We might as well inline this `aux2`.
```suggestion
/-- Since we assumed `Q` being Baer, the linear map `x ↦ f' (x • y) : I ⟶ Q` extends to `R ⟶ Q`,
call this extended map `φ`-/
def extension_of_max_adjoin.extend_ideal_to (h : module.Baer R Q) (y : N) : R →ₗ[R] Q :=
(h (extension_of_max_adjoin.ideal hi f y) (extension_of_max_adjoin.ideal_to hi f y)).some

lemma extension_of_max_adjoin.extend_ideal_to_is_extension (h : module.Baer R Q) (y : N) :
  ∀ (x : R) (mem : x ∈ extension_of_max_adjoin.ideal hi f y),
  extension_of_max_adjoin.extend_ideal_to hi f h y x =
  extension_of_max_adjoin.ideal_to hi f y ⟨x, mem⟩ :=
(h (extension_of_max_adjoin.ideal hi f y) (extension_of_max_adjoin.ideal_to hi f y)).some_spec
```